### PR TITLE
Visual Studio 2017

### PR DESCRIPTION
**Pull request by madebr**
Hey,

These patches allow me to build OpenRW using Visual Studio 2017 on Windows 10.

Description:
- ~~cmake: Add LINK_STATIC switch: openal-soft and boost need special defines to know that library is linked as a static library. This eases distribution and testing as no extra dll is needed.~~
- Fix a bunch of compile errors

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -112,24 +112,25 @@ add_subdirectory(rwlib)
...
+            DESTINATION "${DOC_DIR}"
+    )
 endif()
+
 # And copy to build directory for CI
 file(COPY COPYING
-	 DESTINATION .
-	 )
+     DESTINATION .
+     )
```
Comment (JayFoxRox) [inline CMakeLists.txt]: this still has some extra spaces

File: .gitignore [hunk 2]
```diff
@@ -4,4 +4,6 @@ documentation/
 *~
 *swp
 *.user*
+.vs
+*.bak
```
Comment (JayFoxRox) [inline .gitignore]: Fair enough for VS, but in general users should configure this in their global local git config. We can't add settings for every IDE in existence to a project

File: rwengine/src/engine/GameData.cpp [hunk 3]
```diff
@@ -446,6 +446,7 @@ void GameData::loadModel(ModelID model) {
             slotname = name;
             break;
         case ModelDataType::PedInfo:
+        {
```
Comment (JayFoxRox) [inline rwengine/src/engine/GameData.cpp]: "How did this ever work?" - normally my clang complains about such things.
Also I'd suggest moving the opening brace on the previous line as we do for `if` and functions. But it's ultimately @tsjost and @danhedron call

File: cmake/modules/FindGLM.cmake [hunk 4]
```diff
@@ -56,10 +56,6 @@ INCLUDE(FindPackageHandleStandardArgs)
...
 
-IF(GLM_FOUND)
-    SET(GLM_INCLUDE_DIRS "${GLM_INCLUDE_DIR}")
+SET(GLM_INCLUDE_DIRS "${GLM_INCLUDE_DIR}")
 
-    IF(NOT GLM_FIND_QUIETLY)
-        MESSAGE(STATUS "GLM_INCLUDE_DIR = ${GLM_INCLUDE_DIR}")
-    ENDIF(NOT GLM_FIND_QUIETLY)
-ENDIF(GLM_FOUND)
+mark_as_advanced(GLM_INCLUDE_DIR)
```
Comment (JayFoxRox) [inline cmake/modules/FindGLM.cmake]: Did you write this patch yourself or did you look for an updated FindGLM?

File: cmake/modules/FindLibSndFile.cmake [hunk 5]
```diff
@@ -19,16 +19,14 @@ find_path(LIBSNDFILE_INCLUDE_DIR
...
-find_package_handle_standard_args(LibSndFile  DEFAULT_MSG  LIBSNDFILE_LIBRARY LIBSNDFILE_INCLUDE_DIR)
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(LibSndFile DEFAULT_MSG LIBSNDFILE_LIBRARY LIBSNDFILE_INCLUDE_DIR)
 
-if(LIBSNDFILE_FOUND)
-  set(LIBSNDFILE_LIBRARIES ${LIBSNDFILE_LIBRARY})
-  set(LIBSNDFILE_INCLUDE_DIRS ${LIBSNDFILE_INCLUDE_DIR})
-endif(LIBSNDFILE_FOUND)
+set(LIBSNDFILE_LIBRARIES ${LIBSNDFILE_LIBRARY})
+set(LIBSNDFILE_INCLUDE_DIRS ${LIBSNDFILE_INCLUDE_DIR})
```
Comment (JayFoxRox) [inline cmake/modules/FindLibSndFile.cmake]: Why unconditionally? imo this should still depend on the library being found - that's kind of cmake standard / expected behaviour?

File: tests/test_globals.hpp [hunk 6]
```diff
@@ -31,10 +31,11 @@ struct print_log_value<glm::vec3> {
         s << glm::to_string(v);
     }
 };
+BOOST_NS_MAGIC_CLOSING
 }
 }
-BOOST_NS_MAGIC_CLOSING
 
+#if BOOST_VERSION < 106400
```
Comment (JayFoxRox) [inline tests/test_globals.hpp]: We should add an issue about this after merge so we can remove such checks in a couple of years when all distros upgraded boost.

File: rwengine/src/audio/MADStream.hpp [hunk 7]
```diff
@@ -10,7 +12,13 @@
 #include <stdio.h>
 #include <sys/mman.h>
 #include <sys/stat.h>
+#if defined(HAVE_UNISTD_H)
 #include <unistd.h>
+#elif defined(HAVE_IO_H)
+#include <io.h>
+#else
+#error "Need an alternative for unistd.h and io.h"
+#endif
```
Comment (JayFoxRox) [inline rwengine/src/audio/MADStream.hpp]: imo these 2 should be implied for unix like platforms and seperate checks / implementations should be added for windows imo

File: tests/CMakeLists.txt [hunk 8]
```diff
@@ -53,10 +53,14 @@ set(TEST_SOURCES
...
-ADD_DEFINITIONS(-DBOOST_TEST_DYN_LINK)
-
 add_executable(run_tests ${TEST_SOURCES})
 
+if(NOT LINK_STATIC)
+	target_compile_definitions(run_tests
+		PUBLIC
+			BOOST_TEST_DYN_LINK)
+endif()
+
```
Comment (JayFoxRox) [inline tests/CMakeLists.txt]: I feel like this should be a seperate PR.

I feel we should discuss what the standard cmake way of enforcing this is, I'd assume there are ways of doing this?

File: CMakeLists.txt [hunk 9]
```diff
@@ -7,10 +7,12 @@ project(OpenRW)
 #
 
 # Global Build Configuration
-set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DRW_DEBUG=1")
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14 -Wall -pthread -Wextra -Wpedantic")
-set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/modules")
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
 
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DRW_DEBUG=1")
+if(NOT MSVC)
```
Comment (JayFoxRox) [inline CMakeLists.txt]: this check should be the other way around imo. Add compiler specific flags, not avoiding incompatible flags

File: CMakeLists.txt [hunk 10]
```diff
@@ -108,14 +108,6 @@ find_package(GLM REQUIRED)
 find_package(LibSndFile REQUIRED)
 find_package(SDL2 REQUIRED)
 
-include_directories(
-    ${SDL2_INCLUDE_DIR}
-    ${GLM_INCLUDE_DIRS}
-    ${OPENGL_INCLUDE_DIR}
-    ${BULLET_INCLUDE_DIRS}
-    ${OPENAL_INCLUDE_DIR}
-)
```
Comment (JayFoxRox) [inline CMakeLists.txt]: This commit also feels like more build changes which are not related to VS2017, they should probably be a seperate PR

File: cmake/modules/FindGLM.cmake (hunk 4 above)
Comment (madebr) [inline cmake/modules/FindGLM.cmake]: I wrote it myself.

Comment (JayFoxRox) [inline cmake/modules/FindGLM.cmake]: I suggest using a file you find elsewhere for this. GLM is a pretty common lib. CMake depends on standardization. Often cmake files are promoted to being included in newer CMake versions.

Also note that GLM used to come with it's own cmake file apparently (which we should probably use as it's the most likely to be included in CMake upstream): https://chromium.googlesource.com/external/github.com/g-truc/glm/+/0.9.0/FindGLM.cmake (not sure if there are more recent versions or if/why it was removed)

File: tests/CMakeLists.txt (hunk 8 above)
Comment (madebr) [inline tests/CMakeLists.txt]: I will change it to this: https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html

File: .gitignore (hunk 2 above)
Comment (madebr) [inline .gitignore]: Changed it to only ignore bak files. Those are artifacts of WinMerge.

File: tests/CMakeLists.txt (hunk 8 above)
Comment (madebr) [inline tests/CMakeLists.txt]: On second thought, I believe the current behavior is fine. The naming might be off. BUILD_SHARED_LIBS changes the behavior of add_library, which is not wanted.

File: rwengine/src/engine/GameData.cpp (hunk 3 above)
Comment (madebr) [inline rwengine/src/engine/GameData.cpp]: Moved the brace.
I haven't checked but maybe it 'worked' because the variable is static?

File: cmake/modules/FindGLM.cmake (hunk 4 above)
Comment (madebr) [inline cmake/modules/FindGLM.cmake]: Retested and the patches to findLibSndFile and findGLM are not needed.
So left them out.

File: tests/test_globals.hpp (hunk 6 above)
Comment (madebr) [inline tests/test_globals.hpp]: I'm not surewhich version added the nullptr_t specialization. It might be 1.63.0 (or lower).

File: cmake/modules/FindLibSndFile.cmake (hunk 5 above)
Comment (madebr) [inline cmake/modules/FindLibSndFile.cmake]: I kicked the FindLibSndFile.cmake patch out of my patch set.

File: CMakeLists.txt (hunk 10 above)
Comment (madebr) [inline CMakeLists.txt]: Ok, will do another PR for that.

File: tests/test_globals.hpp (hunk 6 above)
Comment (JayFoxRox) [inline tests/test_globals.hpp]: Please check and fix this check or mark this comment as a todo. I'd go as far and call this a bug otherwise ;)

File: CMakeLists.txt (hunk 9 above)
Comment (madebr) [inline CMakeLists.txt]: Changed it

File: rwengine/src/script/ScriptModule.hpp [hunk 11]
```diff
@@ -175,9 +175,11 @@ class ScriptModule {
     std::map<ScriptFunctionID, ScriptFunctionMeta> functions;
 };
 
+#if !defined(_MSC_VER)
 template <>
 void ScriptModule::bind<void(const ScriptArguments&)>(
     ScriptFunctionID id, int args, void (*function)(const ScriptArguments&));
+#endif
```
Comment (JayFoxRox) [inline rwengine/src/script/ScriptModule.hpp]: If this is not necessary for MSVC, why is this necessary in the first place?
Unless there's a compiler bug / extension I'd assume such things to be the same across compilers with the same C++ std.

Comment (madebr) [inline rwengine/src/script/ScriptModule.hpp]: It can also be commented out on gcc 6.3.1 and clang 3.8.1.
Since we are using -std=c++14, I can assume all supported compilers do not need this specialization.

Comment (madebr) [inline rwengine/src/script/ScriptModule.hpp]: The specialization has now been removed.

File: tests/CMakeLists.txt (hunk 8 above)
Comment (madebr) [inline tests/CMakeLists.txt]: Removed this patch. Will open a new PR/issue about this when this gets merged.

File: tests/test_globals.hpp (hunk 6 above)
Comment (madebr) [inline tests/test_globals.hpp]: Added a comment: specialization needed in 1.60.0, not needed in 1.64.0.

Comment (madebr) [inline tests/test_globals.hpp]: using [wandbox.org](https://wandbox.org), I've verified that the `nullptr_t` specialization is included in boost starting from version 1.64.0.
1.63.0 needs the specialization: https://wandbox.org/permlink/XL4uPZWetPGcMbT9
1.64.0 fails to compile: https://wandbox.org/permlink/F8qO7hUFbX3C4ITE

**Comment by madebr** [comment]
Added AppVeyor Windows CI:
- I wrote the build logic in a python script `build_windows.py`. That way it is possible to run and develop it on my own system.
- The Windows dependencies are available on https://github.com/madebr/openrw-windows-dependencies. You might want to clone that repository to the openrw team github account and modify `WIN_DEP_URL` in `build_windows.py`.
- example output: https://ci.appveyor.com/project/madebr/openrw/build/1.0.72
- Artifacts are available: e.g. https://ci.appveyor.com/project/madebr/openrw/build/job/6uu613fx93nlju91/artifacts

@danhedron @JayFoxRox : _final_ review requested

**Comment by madebr** [comment]
- Rebased on master

New features:
- Updated `FindFFmpeg.cmake` to work on Windows
- Including `Windows.h` in `defines.h` in PR rwengine/openrw#313 included lots of things. Avoid that.
- Do not push artifacts by default: this behavior can be overridden by including `[artifact:1]` in your commit message.

**Comment by darkf** [comment]
- What's wrong with using SDLmain?
- It's not clear how you changed FindFFmpeg because you changed all of the formatting in the same commit...
- You seem to be using "Windows" to mean "Visual Studio"

**Comment by madebr** [comment]
Thanks for the review:
- I added the call to `SDL_SetMainReady` to the tests first because the Boost framework does not need a main function as entry point. At the same time, I added the same logic to rwgame. I wanted the differences between the game and tests to be small.
- The FindFFmpeg commit _can_ be separated in multiple patches. I don't see the added value in dissecting this commit. The commit message contains the list of changes:
  * use FindPackageHandleStandardArgs package
  * add swresample on Windows (libavcodec has a dependency on swresample)
  * additionally link to secur32.lib and Ws2_32.lib
  * formatting changes..
- Sorry about the slip of the tongue about Visual Studio/Windows. This PR is about adding Visual Studio support. Do you refer to the naming of the build script `build_windows.py`?

**Comment by madebr** [comment]
- first patches fix compilation errors on windows
- Simplified the CMake module overrides and fix a link error on windows.
- rwgame needs a main function, so define `SDL_MAIN_HANDLED` in `main.cpp`.
- I've unified the testing for windows and linux:
  * `cmake/ctest/build.ctest` configures, builds, tests, does coverage, ... and submits
  * `cmake/ctest/calculate_names.ctest` creates names for the build (for CDash)
  * The default configuration for a system can be found in `cmake/ctest/configure_{LINUX/WINDOWS].ctest`
  * For CI, some extra environmental variables must be set. see `.travis.yml` and `.appveyor.yml`
  * For the experimental script, run it as `ctest -S cmake/ctest/script_experimental.ctest -DRUN_COVERAGE=TRUE`

**Comment by madebr** [comment]
Actually all patches (except the spelling fix :D ) are needed for building on Visual Studio.
- The simplification of the packages is needed as well, since CMake on Windows thought the glu32 dependency of OpenGL was an object instead of a library.
- The last patch is strictly not needed but is for AppVeyor.

What about [the Windows dependencies](https://github.com/madebr/openrw-windows-dependencies). Should it stay there?

**Comment by cornytrace** [comment]
What is the progress on merging this?

**Comment by madebr** [comment]
Updated the windows dependencies and this patch series:
- Result at https://ci.appveyor.com/project/madebr/openrw/build/1.0.376

Things left to do or to discuss:
- where to put [openrw-windows-dependencies](https://github.com/madebr/openrw-windows-dependencies)?
- create a nice how-to + script for windows users to fetch the dependencies and start developing
- create an account op appveyor.com, set-up CI and add a badge to the readme
